### DISCUSSION TITLE
Resource pack + tag

### DISCUSSION THREAD

**Discussion by DevBlokJochem** [Minestom/Minestom#1307] [Help] [answered]
Hello, I want to create some custom item with optifine if needed. The problem is that I dont know how to do this. So my situation is: I want to create 2 new swords with the material DIAMOND_SWORD. 
I want to know:
-which tag do I have to apply?
-which file do I have to add in my resourcepack next to the image?

**Comment by MrGazdag** [comment, marked as answer]
Not sure why you would need Optifine for this, you can just use regular CustomModelData, thats already present in vanilla and thats probably what you want. I will detail an example, where we want to make a fire sword and an ice sword.

### Minestom
You can set the CustomModelData value on an ItemMeta with `ItemMeta.Builder#customModelData(int)`. Using this, to give a player the custom swords, you can use the following code:
```java
Player player = ...;

// Construct the items
ItemStack diamondSword = ItemStack.of(Material.DIAMOND_SWORD);
ItemStack fireSword = ItemStack.builder(Material.DIAMOND_SWORD).meta(meta -> meta.customModelData(1)).build();
ItemStack iceSword = ItemStack.builder(Material.DIAMOND_SWORD).meta(meta -> meta.customModelData(2)).build();

// Give the player the items
player.getInventory().addItemStack(diamondSword);
player.getInventory().addItemStack(fireSword);
player.getInventory().addItemStack(iceSword);
```

### Resource Pack
For the resourcepack part, you need to modify the existing `assets/minecraft/models/item/diamond_sword.json` file, to include overrides for the custom swords. In this example, the textures for the two swords are located in:
- `assets/minecraft/textures/item/fire_sword_texture.png`
- `assets/minecraft/textures/item/ice_sword_texture.png`

You have to make two model files for the two swords. `assets/minecraft/models/item/fire_sword_model.json`:
```json
{
    "parent": "item/handheld",
    "textures": {
        "layer0": "item/fire_sword_texture"
    }
}
```
And, similarly, `assets/minecraft/models/item/ice_sword_model.json`:
```json
{
    "parent": "item/handheld",
    "textures": {
        "layer0": "item/ice_sword_texture"
    }
}
```

Now, in order for these to appear when a diamond sword has CustomModelData, you need to make two overrides in the `diamond_sword.json` file, like this:
```json
{
    "parent": "item/handheld",
    "textures": {
        "layer0": "item/diamond_sword"
    },
    "overrides": [
        {"predicate":{"custom_model_data":1},"model":"item/fire_sword_model"},
        {"predicate":{"custom_model_data":2},"model":"item/ice_sword_model"}
    ]
}
```

Now, you should be able to see three different swords, depending on the CustomModelData:
- a value of 0 (or not specified) results in the default diamond sword
- a value of 1 results in the fire sword
- a value of 2 (or more) results in the ice sword.

*Full disclosure: i have not tested the model code here, if it does not work, let me know.*

**Comment by DevBlokJochem** [comment]
This is all without optine or something. You can create custom models for entitys too, but not in vanilla. IAM has a good solution for this. Look at this repo to implement custom entitys: https://github.com/WorldSeedGames/WorldSeedEntityEngine . There is also a demo server in the test folder.
